File: java/src/test/java/fyi/oxide/pdf/PdfPolicyTest.java
/*
 * Copyright 2025-2026 Yury Fedoseev and pdf_oxide contributors.
 * Licensed under MIT OR Apache-2.0.
 */
package fyi.oxide.pdf;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import fyi.oxide.pdf.exception.PdfException;
import fyi.oxide.pdf.policy.PolicyMode;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;

/**
 * Tests the global crypto-governance policy (v0.3.50 #230). pdf_oxide
 * is **set-once**: a single {@link PdfPolicy#set(PolicyMode)} call at
 * process startup, before any other crypto operation, is permitted.
 * Subsequent {@code set} calls throw. The default lazy initialisation
 * (any first {@link PdfPolicy#current()} or other crypto access) seeds
 * the policy to {@link PolicyMode#COMPAT}.
 *
 * <p>Surefire is configured with {@code reuseForks=false}, so each
 * test class gets a fresh JVM. We use {@code @Order} within this
 * class to make sure the {@code set()} attempt runs BEFORE any
 * {@code current()} read that would lazily lock the policy.
 */
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class PdfPolicyTest {

    /**
     * Run FIRST in this JVM fork: this is the only safe place to
     * call {@code set()} before another test's {@code current()}
     * lazily initialises the policy to COMPAT.
     */
    @Test
    @Order(1)
    void setSwitchesToStrictAtProcessStart() {
        PdfPolicy.set(PolicyMode.STRICT);
        assertThat(PdfPolicy.current()).isEqualTo(PolicyMode.STRICT);
    }

    @Test
    @Order(2)
    void secondSetThrowsAlreadySet() {
        // The previous test set the policy to STRICT. Any further
        // set() call should fail with the set-once error.
        assertThatThrownBy(() -> PdfPolicy.set(PolicyMode.COMPAT))
                .isInstanceOf(PdfException.class)
                .hasMessageContaining("already set");
    }

    @Test
    @Order(3)
    void presetAccessorsReturnTheRightMode() {
        // Read-only — independent of process state.
        assertThat(PdfPolicy.compat()).isEqualTo(PolicyMode.COMPAT);
        assertThat(PdfPolicy.strict()).isEqualTo(PolicyMode.STRICT);
        assertThat(PdfPolicy.fipsStrict()).isEqualTo(PolicyMode.FIPS_STRICT);
    }
}

File: php/tests/Integration/PdfPolicyTest.php
<?php

/*
 * Copyright 2025-2026 Yury Fedoseev and pdf_oxide contributors.
 * Licensed under MIT OR Apache-2.0.
 */

declare(strict_types=1);

namespace PdfOxide\Tests\Integration;

use PdfOxide\PdfPolicy;

/**
 * Smoke tests for {@see PdfPolicy}, mirroring
 * `java/src/test/java/fyi/oxide/pdf/PdfPolicyTest.java`.
 *
 * pdf_oxide is set-once per process — we DON'T call `set()` here so
 * we don't lock the policy for the rest of the PHPUnit run. Other
 * integration tests open documents (which may lazily seed the policy
 * to `compat`), so we only test read-only accessors + preset returns.
 */
final class PdfPolicyTest extends IntegrationTestCase
{
    public function testCurrentReturnsString(): void
    {
        $current = PdfPolicy::current();
        $this->assertIsString($current);
        $this->assertNotEmpty($current);
    }

    public function testPresetAccessors(): void
    {
        $this->assertSame('compat', PdfPolicy::compat());
        $this->assertSame('strict', PdfPolicy::strict());
        $this->assertSame('fips_strict', PdfPolicy::fipsStrict());
    }

    public function testFipsAvailableReturnsBool(): void
    {
        $this->assertIsBool(PdfPolicy::fipsAvailable());
    }

    public function testActiveProviderReturnsString(): void
    {
        $this->assertIsString(PdfPolicy::activeProvider());
    }
}
